src/lib.rs

```diff
@@ -5,651 +5,741 @@
-    fn heap_size_of_children(&self) -> usize {
-        self.0.heap_size_of_children() +
-        self.1.heap_size_of_children() +
-        self.2.heap_size_of_children()
-    }
+/// In most cases, `Name`s are compared ignoring case.
+#[derive(Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub struct Name<'a> {
+    // TODO: optimize with an Atom-like thing
+    // There a `const` Names, and so it is possible for the statis strings
+    // to havea different memory address. Additionally, when used in match
+    // statements, the strings are compared with a memcmp, possibly even
+    // if the address and length are the same.
+    //
+    // Being an enum with an Atom variant that is a usize (and without a
+    // string pointer and boolean) would allow for faster comparisons.
+    source: &'a str,
+    insensitive: bool,
 }
 
-impl<LHS: AsRef<[Param]>, RHS: AsRef<[Param]>> PartialEq<Mime<RHS>> for Mime<LHS> {
-    #[inline]
-    fn eq(&self, other: &Mime<RHS>) -> bool {
-        self.0 == other.0 && self.1 == other.1 && self.2.as_ref() == other.2.as_ref()
-    }
+/// An error when parsing a `Mime` from a string.
+#[derive(Debug)]
+pub struct FromStrError {
+    inner: parse::ParseError,
 }
 
-/// Easily create a Mime without having to import so many enums.
-///
-/// # Example
-///
-/// ```
-/// # #[macro_use] extern crate mime;
-///
-/// # fn main() {
-/// let json = mime!(Application/Json);
-/// let plain = mime!(Text/Plain; Charset=Utf8);
-/// let text = mime!(Text/Html; Charset=("bar"), ("baz")=("quux"));
-/// let img = mime!(Image/_);
-/// # }
-/// ```
-#[macro_export]
-macro_rules! mime {
-    ($top:tt / $sub:tt) => (
-        mime!($top / $sub;)
-    );
-
-    ($top:tt / $sub:tt ; $($attr:tt = $val:tt),*) => (
-        $crate::Mime(
-            __mime__ident_or_ext!(TopLevel::$top),
-            __mime__ident_or_ext!(SubLevel::$sub),
-            vec![ $((__mime__ident_or_ext!(Attr::$attr), __mime__ident_or_ext!(Value::$val))),* ]
-        )
-    );
+#[derive(Clone)]
+enum Source {
+    Atom(u8, &'static str),
+    Dynamic(String),
 }
 
-#[doc(hidden)]
-#[macro_export]
-macro_rules! __mime__ident_or_ext {
-    ($enoom:ident::_) => (
-        $crate::$enoom::Star
-    );
-    ($enoom:ident::($inner:expr)) => (
-        $crate::$enoom::Ext($inner.to_string())
-    );
-    ($enoom:ident::$var:ident) => (
-        $crate::$enoom::$var
-    )
+impl Source {
+    fn as_ref(&self) -> &str {
+        match *self {
+            Source::Atom(_, s) => s,
+            Source::Dynamic(ref s) => s,
+        }
+    }
 }
 
-macro_rules! enoom {
-    (pub enum $en:ident; $ext:ident; $($ty:ident, $text:expr;)*) => (
+#[derive(Clone)]
+enum Params {
+    Utf8(usize),
+    Custom(usize, Vec<(Str, Str)>),
+    None,
+}
 
-        #[derive(Clone, Debug, Eq, Hash, Ord, PartialOrd)]
-        pub enum $en {
-            $($ty),*,
-            $ext(String)
+#[derive(Clone, Copy)]
+struct Str(usize, usize);
```

Comment:
I agree, the name here is not good. I think I'll switch to `Indices` or `Indexed`.